### ISSUE TITLE
'when' condition swallowing error on 'is defined' check

### ISSUE THREAD

**Issue by feanil** [ansible/ansible#28733]
##### ISSUE TYPE

 - Bug Report

##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file = 
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

Running from and to Ubuntu 16.10

##### SUMMARY

The problem occurs when we use the `when` condition to check the definition of a variable.  If the variable value resolution has an error, it is swallowed as if the variable is not defined.
###### playbook.yml
```yaml
- hosts: all
  tasks:
    - name: shellout
      shell: "echo {{ config }}"
      when: config is defined
```
###### extra-vars.yml
```yaml
config: "{{ BAD_VAR }}"
```

1. Put the above two files into a new empty folder.
2. Run the playbook passing in the extra args.  `ansible-playbook playbook.yml -c local -i localhost, -e@extra-vars.yml  -vvv`

##### EXPECTED RESULTS

The command should have failed and told me that `BAD_VAR` is not defined.(This is what happens if instead of checking for `is defined` you just check for truthiness. eg. `when: config`

```
TASK [shellout] ****************************************************************
task path: /home/feanil/test/ansible/playbook.yml:3
fatal: [localhost]: FAILED! => {
    "failed": true, 
    "msg": "The conditional check 'config' failed. The error was: error while evaluating conditional (config): 'BAD_VAR' is undefined\n\nThe error appears to have been in '/home/feanil/test/ansible/playbook.yml'
: line 3, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - name: shellout\n      ^ here\n"
}
```
##### ACTUAL RESULTS

```
TASK [shellout] ****************************************************************
task path: /home/feanil/test/ansible/playbook.yml:3
skipping: [localhost] => {
    "changed": false, 
    "skip_reason": "Conditional check failed", 
    "skipped": true
}
```

**Comment by samdoran** [comment]
I believe this behavior is correct: `config` _is_ defined, therefore the task runs. `BAD_VAR` is undefined, causing a failure. Ansible reports that the `BAD_VAR` is undefined, which you listed in your expected results:

> The command should have failed and told me that BAD_VAR is not defined.

This is most likely related to issue #24338 and the user of `StrictUndefined` for variables.

needs_info

**Comment by feanil** [comment]
@samdoran I don't understand your response.  `config` is always defined in my example and so the task should never be skipped.  However, currently it is skipped if the value of `config` references vars that don't exist masking an error.

**Comment by sivel** [comment]
This issue was resolved in the v2.2.2 release of ansible.  From your version output you are running v2.2.0.

Please upgrade to a newer version and you should be all set:

```
ansible --version
ansible 2.2.2.0 (detached HEAD 31300bd36b) last updated 2017/08/30 19:34:27 (GMT -500)
```

```
TASK [shellout] ****************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'config is defined' failed. The error was: error while evaluating conditional (config is defined): {{ BAD_VAR }}: 'BAD_VAR' is undefined\n\nThe error appears to have been in '/Users/matt/projects/ansibledev/playbooks/28733.yml': line 3, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - name: shellout\n      ^ here\n"}
```

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by samdoran** [comment]
@feanil Looks like I stared at it too long and flipped "Expected" with "Actual" in your report. Sorry for the confusion, but this has been fixed in 2.2.2 as @sivel mentioned. Please upgrade to a newer version and that should resolve your issue.
